File: src/ui/components/code_block.py
import gi
gi.require_version('Gtk', '4.0')
from gi.repository import Gtk, Gdk, GLib

class CodeBlock(Gtk.Box):
    """A widget for displaying code blocks with a copy button."""
    
    def __init__(self, content: str, language: str = "text", *args, **kwargs):
        super().__init__(orientation=Gtk.Orientation.VERTICAL, *args, **kwargs)
        self.add_css_class("code-block")
        
        # Header
        header = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL)
        header.add_css_class("code-header")
        
        lang_label = Gtk.Label(label=language.upper())
        lang_label.add_css_class("code-lang-label")
        lang_label.set_hexpand(True)
        lang_label.set_halign(Gtk.Align.START)
        header.append(lang_label)
        
        copy_btn = Gtk.Button()
        copy_btn.set_icon_name("edit-copy-symbolic")
        copy_btn.add_css_class("flat")
        copy_btn.add_css_class("copy-button")
        copy_btn.set_tooltip_text("Copy")
        copy_btn.connect("clicked", self._on_copy_clicked)
        header.append(copy_btn)
        
        self.append(header)
        
        # Content
        self.content = content
        
        scrolled = Gtk.ScrolledWindow()
        scrolled.set_policy(Gtk.PolicyType.AUTOMATIC, Gtk.PolicyType.NEVER) # Horizontal scroll only
        scrolled.set_max_content_height(400) # Prevents huge blocks from taking over
        
        # Using TextView for better code display (selection, etc)
        text_view = Gtk.TextView()
        text_view.set_editable(False)
        text_view.set_monospace(True)
        text_view.set_wrap_mode(Gtk.WrapMode.NONE) # Code usually better without wrapping
        text_view.add_css_class("code-content")
        
        buffer = text_view.get_buffer()
        buffer.set_text(self.content, -1)
        
        scrolled.set_child(text_view)
        self.append(scrolled)
        
    def _on_copy_clicked(self, button):
        clipboard = Gdk.Display.get_default().get_clipboard()
        clipboard.set(self.content)
        
        # Visual feedback
        icon = button.get_child() 
        # (Usually an Adw.ButtonContent or Image, but simpler for standard icon button)
        
        # Change icon temporarily
        button.set_icon_name("object-select-symbolic")
        
        def restore():
            button.set_icon_name("edit-copy-symbolic")
            return False
            
        GLib.timeout_add(1000, restore)

File: src/ui/components/code_block.rs
use gtk4::prelude::*;
use gtk4::{gdk, glib};

pub struct CodeBlock {
    pub widget: gtk4::Box,
}

impl CodeBlock {
    pub fn new(language: &str, code: &str) -> Self {
        let widget = gtk4::Box::new(gtk4::Orientation::Vertical, 0);
        widget.add_css_class("code-block");

        let header = gtk4::Box::new(gtk4::Orientation::Horizontal, 0);
        header.add_css_class("code-header");

        let lang_label = gtk4::Label::new(Some(&language.to_uppercase()));
        lang_label.add_css_class("code-lang-label");
        lang_label.set_hexpand(true);
        lang_label.set_halign(gtk4::Align::Start);
        header.append(&lang_label);

        let content = code.to_string();
        let copy_btn = gtk4::Button::builder()
            .icon_name("edit-copy-symbolic")
            .css_classes(["flat", "copy-button"])
            .tooltip_text("Copy")
            .build();
        let content = content.to_string();
        copy_btn.connect_clicked(move |btn| {
            if let Some(display) = gdk::Display::default() {
                let clipboard = display.clipboard();
                clipboard.set_text(&content);
            }
            btn.set_icon_name("object-select-symbolic");
            let btn_clone = btn.clone();
            glib::timeout_add_local(std::time::Duration::from_millis(1000), move || {
                btn_clone.set_icon_name("edit-copy-symbolic");
                glib::ControlFlow::Break
            });
        });
        header.append(&copy_btn);

        widget.append(&header);

        let scrolled = gtk4::ScrolledWindow::new();
        scrolled.set_policy(gtk4::PolicyType::Automatic, gtk4::PolicyType::Never);
        scrolled.set_max_content_height(400);

        let text_view = gtk4::TextView::new();
        text_view.set_editable(false);
        text_view.set_monospace(true);
        text_view.set_wrap_mode(gtk4::WrapMode::None);
        text_view.add_css_class("code-content");

        let buffer = text_view.buffer();
        buffer.set_text(code);

        scrolled.set_child(Some(&text_view));
        widget.append(&scrolled);

        CodeBlock { widget }
    }
}
